File: iosApp/iosApp/UI/Components/Cards.swift
import SwiftUI

/// TA33 event hero karta - brandová hlavička Teplicko-adršpašského pochodu: datum + odznak „33 KM",
/// wordmark „TA33" (33 oranžově), název akce, místo + start, a silueta adršpašských skalních věží dole.
/// Slate identita + oranžový akcent. Zrcadlí Android `IdentityCard`.
struct IdentityCard: View {
    let date: String
    let place: String
    let sub: String

    var body: some View {
        ZStack(alignment: .bottom) {
            sunriseGlow
            rockRidge

            VStack(alignment: .leading, spacing: Ta33Spacing.x2) {
                HStack(alignment: .center) {
                    Text(date.uppercased())
                        .font(Ta33Font.overline)
                        .kerning(1.3)
                        .foregroundStyle(Ta33Color.fgOnDarkMuted)
                    Spacer(minLength: Ta33Spacing.x3)
                    distanceBadge
                }

                (Text("TA").foregroundStyle(Ta33Color.fgOnDark)
                    + Text("33").foregroundStyle(Ta33Color.orange))
                    .font(Ta33Font.display1)
                    .kerning(0.5)

                Text("Teplicko-adršpašský pochod".uppercased())
                    .font(Ta33Font.caption)
                    .kerning(1.6)
                    .foregroundStyle(Ta33Color.fgOnDarkMuted)

                VStack(alignment: .leading, spacing: Ta33Spacing.x1) {
                    Text(place)
                        .font(Ta33Font.h3)
                        .foregroundStyle(Ta33Color.fgOnDark)
                    Text(sub)
                        .font(Ta33Font.small)
                        .foregroundStyle(Ta33Color.fgOnDarkMuted)
                }
                .padding(.top, Ta33Spacing.x2)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(Ta33Spacing.x6)
            .padding(.bottom, Ta33Spacing.x9) // místo pro horizont skalních věží
        }
        .background(
            LinearGradient(
                colors: [Ta33Color.slate900, Ta33Color.slate800],
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )
        )
        .clipShape(RoundedRectangle(cornerRadius: Ta33Radius.xl, style: .continuous))
        .shadow(color: .black.opacity(0.12), radius: Ta33Spacing.x2, x: 0, y: Ta33Spacing.x1)
    }

    private var distanceBadge: some View {
        Text("33 KM")
            .font(Ta33Font.caption)
            .kerning(0.8)
            .foregroundStyle(Ta33Color.fgOnOrange)
            .padding(.horizontal, Ta33Spacing.x3)
            .padding(.vertical, Ta33Spacing.x1)
            .background(Ta33Color.orange, in: Capsule())
    }

    /// Jemný oranžový „východ slunce" za skalními věžemi (dawn glow u horizontu).
    private var sunriseGlow: some View {
        RadialGradient(
            colors: [Ta33Color.orange.opacity(0.28), Ta33Color.orange.opacity(0.0)],
            center: .bottom,
            startRadius: 0,
            endRadius: 190
        )
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .bottom)
        .blendMode(.plusLighter)
        .allowsHitTesting(false)
    }

    /// Nepravidelná silueta skalních věží (Adršpašsko-teplické skály) podél spodní hrany karty.
    private var rockRidge: some View {
        RockRidge()
            .fill(Ta33Color.slate700)
            .frame(maxWidth: .infinity)
            .frame(height: Ta33Spacing.x10) // ~72pt vysoký horizont
    }
}

/// Silueta shluku pískovcových věží: úzké nepravidelné pilíře různých výšek a šířek sdílející základnu.
private struct RockRidge: Shape {
    /// (šířkový poměr, výškový poměr 0..1) pro každou věž zleva doprava.
    private let towers: [(w: CGFloat, h: CGFloat)] = [
        (1.0, 0.42), (0.65, 0.68), (1.35, 0.52), (0.55, 0.9), (0.9, 0.64),
        (1.15, 0.46), (0.5, 0.82), (1.45, 0.6), (0.8, 1.0), (0.6, 0.55),
        (1.2, 0.74), (0.7, 0.48), (1.0, 0.86), (0.85, 0.58), (1.3, 0.7),
    ]

    func path(in rect: CGRect) -> Path {
        var p = Path()
        let gap: CGFloat = 2.5
        let totalWeight = towers.reduce(0) { $0 + $1.w }
        let usableW = rect.width - gap * CGFloat(towers.count - 1)
        let scale = usableW / totalWeight

        p.move(to: CGPoint(x: rect.minX, y: rect.maxY))
        var x = rect.minX
        for (i, t) in towers.enumerated() {
            let w = t.w * scale
            let topY = rect.maxY - rect.height * t.h
            let cap = min(w * 0.5, 5)
            p.addLine(to: CGPoint(x: x, y: topY + cap))
            p.addQuadCurve(to: CGPoint(x: x + cap, y: topY), control: CGPoint(x: x, y: topY))
            p.addLine(to: CGPoint(x: x + w - cap, y: topY))
            p.addQuadCurve(to: CGPoint(x: x + w, y: topY + cap), control: CGPoint(x: x + w, y: topY))
            p.addLine(to: CGPoint(x: x + w, y: rect.maxY))
            x += w
            if i < towers.count - 1 {
                p.addLine(to: CGPoint(x: x + gap, y: rect.maxY)) // úzká štěrbina k základně
                x += gap
            }
        }
        p.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY))
        p.closeSubpath()
        return p
    }
}

/// Bílá „paper" karta - univerzální povrch pro operační obsah. Zrcadlí Android `PaperCard`.
struct PaperCard<Content: View>: View {
    @ViewBuilder let content: Content

    var body: some View {
        content
            .padding(Ta33Spacing.x4)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Ta33Color.paper)
            .clipShape(RoundedRectangle(cornerRadius: Ta33Radius.lg, style: .continuous))
            .shadow(color: .black.opacity(0.08), radius: Ta33Spacing.x2, x: 0, y: Ta33Spacing.x1)
    }
}

#Preview("IdentityCard") {
    IdentityCard(
        date: "Sobota 19. 9. 2026",
        place: "Teplice n. Metují",
        sub: "Start 7:00–10:00 · prezentace u sokolovny"
    )
    .padding(Ta33Spacing.x5)
    .background(Ta33Color.cream)
}

#Preview("PaperCard") {
    PaperCard {
        VStack(alignment: .leading, spacing: Ta33Spacing.x2) {
            Text("Obsah karty").font(Ta33Font.h2).foregroundStyle(Ta33Color.fgStrong)
            Text("Vše operační jde sem.").font(Ta33Font.body).foregroundStyle(Ta33Color.fgMuted)
        }
    }
    .padding(Ta33Spacing.x5)
    .background(Ta33Color.cream)
}

File: androidApp/src/main/kotlin/com/example/ta33/ui/components/Cards.kt
package com.example.ta33.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.example.ta33.ui.theme.Ta33Palette
import com.example.ta33.ui.theme.Ta33Theme

/** Nepravidelné pískovcové věže (Adršpašsko-teplické skály): (šířkový poměr, výškový poměr 0..1). */
private val ROCK_TOWERS = listOf(
    1.0f to 0.42f, 0.65f to 0.68f, 1.35f to 0.52f, 0.55f to 0.9f, 0.9f to 0.64f,
    1.15f to 0.46f, 0.5f to 0.82f, 1.45f to 0.6f, 0.8f to 1.0f, 0.6f to 0.55f,
    1.2f to 0.74f, 0.7f to 0.48f, 1.0f to 0.86f, 0.85f to 0.58f, 1.3f to 0.7f,
)

/**
 * TA33 event hero karta - brandová hlavička Teplicko-adršpašského pochodu: datum + odznak „33 KM",
 * wordmark „TA33" (33 oranžově), název akce, místo + start, a silueta adršpašských skalních věží
 * s jemným oranžovým „východem slunce". Slate identita + oranžový akcent. Zrcadlí iOS `IdentityCard`.
 */
@Composable
fun IdentityCard(
    date: String,
    place: String,
    sub: String,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .fillMaxWidth()
            .shadow(Ta33Theme.spacing.x2, Ta33Theme.radius.xl)
            .clip(Ta33Theme.radius.xl)
            .background(
                Brush.linearGradient(listOf(Ta33Palette.Slate900, Ta33Palette.Slate800)),
            ),
    ) {
        // Skalní silueta + dawn glow podél spodní hrany.
        Canvas(modifier = Modifier.matchParentSize()) {
            val w = size.width
            val h = size.height
            drawRect(
                brush = Brush.radialGradient(
                    colors = listOf(Ta33Palette.Orange.copy(alpha = 0.22f), Color.Transparent),
                    center = Offset(w / 2f, h),
                    radius = w * 0.7f,
                ),
                blendMode = BlendMode.Plus,
            )
            val bandH = 72.dp.toPx()
            val gap = 2.5.dp.toPx()
            val cap = 5.dp.toPx()
            val totalWeight = ROCK_TOWERS.fold(0f) { acc, t -> acc + t.first }
            val usableW = w - gap * (ROCK_TOWERS.size - 1)
            val scale = usableW / totalWeight
            val path = Path().apply {
                moveTo(0f, h)
                var x = 0f
                ROCK_TOWERS.forEachIndexed { i, (wt, ht) ->
                    val tw = wt * scale
                    val topY = h - bandH * ht
                    val c = minOf(tw * 0.5f, cap)
                    lineTo(x, topY + c)
                    quadraticBezierTo(x, topY, x + c, topY)
                    lineTo(x + tw - c, topY)
                    quadraticBezierTo(x + tw, topY, x + tw, topY + c)
                    lineTo(x + tw, h)
                    x += tw
                    if (i < ROCK_TOWERS.size - 1) {
                        lineTo(x + gap, h)
                        x += gap
                    }
                }
                lineTo(w, h)
                close()
            }
            drawPath(path, color = Ta33Palette.Slate700)
        }

        Column(
            modifier = Modifier.padding(
                start = Ta33Theme.spacing.x6,
                top = Ta33Theme.spacing.x6,
                end = Ta33Theme.spacing.x6,
                bottom = Ta33Theme.spacing.x9, // místo pro horizont skalních věží
            ),
            verticalArrangement = Arrangement.spacedBy(Ta33Theme.spacing.x2),
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    text = date.uppercase(),
                    style = MaterialTheme.typography.labelSmall,
                    color = Ta33Theme.colors.fgOnDarkMuted,
                )
                Text(
                    text = "33 KM",
                    style = MaterialTheme.typography.labelMedium,
                    color = Ta33Theme.colors.fgOnOrange,
                    modifier = Modifier
                        .clip(CircleShape)
                        .background(Ta33Palette.Orange)
                        .padding(horizontal = Ta33Theme.spacing.x3, vertical = Ta33Theme.spacing.x1),
                )
            }
            Text(
                text = buildAnnotatedString {
                    withStyle(SpanStyle(color = Ta33Theme.colors.fgOnDark)) { append("TA") }
                    withStyle(SpanStyle(color = Ta33Palette.Orange)) { append("33") }
                },
                style = MaterialTheme.typography.displayLarge,
            )
            Text(
                text = "Teplicko-adršpašský pochod".uppercase(),
                style = MaterialTheme.typography.labelMedium,
                color = Ta33Theme.colors.fgOnDarkMuted,
            )
            Column(verticalArrangement = Arrangement.spacedBy(Ta33Theme.spacing.x1)) {
                Text(
                    text = place,
                    style = MaterialTheme.typography.titleMedium,
                    color = Ta33Theme.colors.fgOnDark,
                )
                Text(
                    text = sub,
                    style = MaterialTheme.typography.bodyMedium,
                    color = Ta33Theme.colors.fgOnDarkMuted,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
    }
}

/**
 * Bílá „paper" karta, radius lg, jemný stín. Kontejner pro operační obsah.
 */
@Composable
fun PaperCard(
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    Surface(
        modifier = modifier
            .fillMaxWidth()
            .shadow(Ta33Theme.spacing.x1, Ta33Theme.radius.lg),
        color = MaterialTheme.colorScheme.surface,
        shape = Ta33Theme.radius.lg,
    ) {
        Column(modifier = Modifier.padding(Ta33Theme.spacing.x4)) {
            content()
        }
    }
}

@Preview
@Composable
private fun IdentityCardPreview() {
    Ta33Theme {
        IdentityCard(
            date = "Sobota 19. 9. 2026",
            place = "Teplice n. Metují",
            sub = "Start 7:00–10:00 · prezentace u sokolovny",
            modifier = Modifier.padding(Ta33Theme.spacing.x5),
        )
    }
}

@Preview
@Composable
private fun PaperCardPreview() {
    Ta33Theme {
        PaperCard(modifier = Modifier.padding(Ta33Theme.spacing.x5)) {
            Text("Obsah karty", style = MaterialTheme.typography.bodyLarge)
        }
    }
}
